app/forms/user_piv_cac_setup_form.rb

```diff
@@ -0,0 +1,91 @@
+class UserPivCacSetupForm
+  extend Forwardable
+
+  State = Struct.new(:x509_dn_uuid, :x509_dn, :error, :success)
+
+  def initialize(user, token)
+    @state = State.new
+    @user = user
+    @token = token
+  end
+
+  def_delegators :@state, :x509_dn_uuid, :x509_dn, :error, :success
+
+  def submit
+    @state.success = valid_token?
+
+    process_valid_submission if success
+
+    FormResponse.new(success: success, errors: {})
+  end
+
+  private
+
+  attr_reader :user, :token
+
+  def process_valid_submission
+    user.x509_dn_uuid = x509_dn_uuid
+    user.save!
+    Event.create(user_id: user.id, event_type: :piv_cac_enabled)
+  rescue PG::UniqueViolation
+    @state.success = false
+    @state.error = 'piv_cac.already_associated'
+  end
+
+  def valid_token?
+    user_has_no_piv_cac &&
+      token_provided &&
+      token_decoded &&
+      not_error_token &&
+      piv_cac_not_already_associated
+  end
+
+  def token_decoded
+    @data = PivCacService.decode_token(@token)
+    if @data
+      true
+    else
+      @state.error = 'token.invalid'
+      false
+    end
+  end
+
+  def not_error_token
+    possible_error = @data['error']
+    if possible_error
+      @state.error = possible_error
+      false
+    else
+      true
+    end
+  end
+
+  def piv_cac_not_already_associated
+    @state.x509_dn_uuid = @data['uuid']
+    @state.x509_dn = @data['dn']
+    if User.find_by(x509_dn_uuid: x509_dn_uuid)
+      @state.error = 'piv_cac.already_associated'
+      false
+    else
+      true
+    end
+  end
+
+  def user_has_no_piv_cac
+    if user.piv_cac_enabled?
+      @state.error = 'user.piv_cac_associated'
+      false
+    else
+      true
+    end
+  end
+
+  def token_provided
```

Comment:
Why not use ActiveModel validations like in our other forms? For example:
```ruby
include ActiveModel::Model

validates :token, presence: true
```

Reply:
I'll see how I can make that work. It does clean things up a bit.